frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/AuthenticationAuthCodeHandler.java

```diff
@@ -0,0 +1,128 @@
+import uk.gov.di.authentication.shared.services.DynamoAuthCodeService;
+import uk.gov.di.authentication.shared.services.DynamoService;
+
+import java.util.Optional;
+
+import static uk.gov.di.authentication.shared.helpers.ApiGatewayResponseHelper.generateApiGatewayProxyErrorResponse;
+import static uk.gov.di.authentication.shared.helpers.ApiGatewayResponseHelper.generateApiGatewayProxyResponse;
+import static uk.gov.di.authentication.shared.helpers.InstrumentationHelper.segmentedFunctionCall;
+import static uk.gov.di.authentication.shared.helpers.LogLineHelper.LogFieldName.AWS_REQUEST_ID;
+import static uk.gov.di.authentication.shared.helpers.LogLineHelper.attachLogFieldToLogs;
+
+public class AuthenticationAuthCodeHandler
+        implements RequestHandler<APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent> {
+
+    private static final Logger LOG = LogManager.getLogger(AuthenticationAuthCodeHandler.class);
+
+    private final Gson gson;
+    private final DynamoAuthCodeService dynamoAuthCodeService;
+    private final ConfigurationService configurationService;
+    private final DynamoService dynamoService;
+    private final AuthorisationCodeService authorisationCodeService;
+
+    public AuthenticationAuthCodeHandler(
+            DynamoAuthCodeService dynamoAuthCodeService,
+            DynamoService dynamoService,
+            ConfigurationService configurationService,
+            AuthorisationCodeService authorisationCodeService) {
+        this.gson = new GsonBuilder().create();
+        this.dynamoAuthCodeService = dynamoAuthCodeService;
+        this.dynamoService = dynamoService;
+        this.configurationService = configurationService;
+        this.authorisationCodeService = authorisationCodeService;
+    }
+
+    public AuthenticationAuthCodeHandler(ConfigurationService configurationService) {
+        this.gson = new GsonBuilder().create();
+        this.configurationService = configurationService;
+        this.dynamoAuthCodeService = new DynamoAuthCodeService(configurationService);
+        this.dynamoService = new DynamoService(configurationService);
+        this.authorisationCodeService = new AuthorisationCodeService(configurationService);
+    }
+
+    @Override
+    public APIGatewayProxyResponseEvent handleRequest(
+            APIGatewayProxyRequestEvent input, Context context) {
+        ThreadContext.clearMap();
+        return segmentedFunctionCall(
+                "auth-code::" + getClass().getSimpleName(),
+                () -> authCodeRequestHandler(input, context));
+    }
+
+    private APIGatewayProxyResponseEvent authCodeRequestHandler(
+            APIGatewayProxyRequestEvent input, Context context) throws Json.JsonException {
+        attachLogFieldToLogs(AWS_REQUEST_ID, context.getAwsRequestId());
+
+        if (configurationService.isAuthCodeHandlerEnabled()) {
+            var authCodeRequest = gson.fromJson(input.getBody(), AuthCodeRequest.class);
```

Comment:
This may be off-base, but we do have an existing Serialisation service class https://github.com/alphagov/di-authentication-api/blob/main/shared/src/main/java/uk/gov/di/authentication/shared/services/SerializationService.java#L23

It also works off Gson, but I wonder if it would give you some freebies in terms of the validation you're doing in this method? 

I _think_ as long as your `AuthCodeRequest` is correctly annoted with `@Required` and `@Expose` and that sort of thing, you might find it would throw you a Json Exception with the missing fields identified?